src/MongoDB.Driver/FilterDefinitionBuilder.cs

```diff
@@ -1541,11 +1551,11 @@ private static void AddClause(BsonDocument document, BsonElement clause)
             else if (document.Contains(clause.Name))
             {
                 var existingClause = document.GetElement(clause.Name);
-                if (existingClause.Value is BsonDocument && clause.Value is BsonDocument)
+                if (existingClause.Value is BsonDocument existingClauseValue && clause.Value is BsonDocument clauseValue)
                 {
-                    var clauseValue = (BsonDocument)clause.Value;
-                    var existingClauseValue = (BsonDocument)existingClause.Value;
-                    if (clauseValue.Names.Any(op => existingClauseValue.Contains(op)))
+                    var topLevelElementName = clauseValue.ElementCount > 0 ? clauseValue.GetElement(0).Name : null;
+                    if (__filtersThatAllowOnlyDollarForm.Contains(topLevelElementName) ||
+                        clauseValue.Names.Any(op => existingClauseValue.Contains(op)))
```

Comment:
Would you be willing to reverse the clauses so that we have the original first and the new one second?

     if (clauseValue.Names.Any(op => existingClauseValue.Contains(op)) ||
         __operatorsThatCantBeCombined.Contains(clauseOperator))